apollo-runtime-kotlin/src/commonTest/kotlin/com/apollographql/apollo/ApolloClientTest.kt

```diff
@@ -0,0 +1,108 @@
+import com.apollographql.apollo.context.DispatchersContext
+import com.apollographql.apollo.mock.MockNetworkTransport
+import com.apollographql.apollo.mock.MockQuery
+import com.apollographql.apollo.mock.TestLoggerExecutor
+import com.apollographql.apollo.network.NetworkResponse
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.retryWhen
+import kotlinx.coroutines.flow.single
+import okio.Buffer
+import okio.ByteString.Companion.encodeUtf8
+import kotlin.test.BeforeTest
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import kotlin.test.assertTrue
+
+@Suppress("EXPERIMENTAL_API_USAGE")
+class ApolloClientTest {
+  private lateinit var networkTransport: MockNetworkTransport
+  private lateinit var apolloClient: ApolloClient
+
+  @BeforeTest
+  fun setUp() {
+    networkTransport = MockNetworkTransport()
+    apolloClient = ApolloClient(
+        networkTransport = networkTransport,
+        executionContext = ExecutionContext.Empty + DispatchersContext(
+            ioDispatcher = Dispatchers.Unconfined
+        ),
+        executors = listOf(TestLoggerExecutor)
+    )
+  }
+
+  @Test
+  fun `when query and success network response, assert success`() {
+    networkTransport.offer(
+        NetworkResponse(
+            body = Buffer().write("{\"data\":{\"name\":\"MockQuery\"}}".encodeUtf8()),
+            executionContext = ExecutionContext.Empty
+        )
+    )
+
+    val response = runBlocking {
+      apolloClient
+          .query(MockQuery())
+          .execute()
+          .single()
+    }
+
+    assertNotNull(response.data)
+    assertEquals(expected = MockQuery.Data, actual = response.data)
+  }
+
+  @Test
+  fun `when query and malformed network response, assert parse error`() {
+    networkTransport.offer(
+        NetworkResponse(
+            body = Buffer(),
+            executionContext = ExecutionContext.Empty
+        )
+    )
+
+    val result = runBlocking {
+      kotlin.runCatching {
+        apolloClient
+            .query(MockQuery())
+            .execute()
+            .single()
+      }
+    }
+
+    assertTrue(result.isFailure)
+    result.onFailure { e ->
+      assertTrue(e is ApolloException)
+      assertTrue(e.error is ApolloError.ParseError)
+    }
+  }
+
+  @Test
+  fun `when query and malformed network response, assert success after retry`() {
+    networkTransport.offer(
+        NetworkResponse(
+            body = Buffer(),
+            executionContext = ExecutionContext.Empty
+        )
+    )
+    networkTransport.offer(
+        NetworkResponse(
+            body = Buffer().write("{\"data\":{\"name\":\"MockQuery\"}}".encodeUtf8()),
+            executionContext = ExecutionContext.Empty
+        )
+    )
+
+    val response = runBlocking {
+      apolloClient
+          .query(MockQuery())
+          .execute()
+          .retryWhen { cause, attempt -> cause is ApolloException && attempt == 0L }
```

Comment:
omg i want the `retryWhen` API for everything I do now

Reply:
yeah nice that we get this out of the box with Flow